vernac/vernacentries.ml

```diff
@@ -2253,7 +2253,30 @@ let locate_if_not_already ?loc (e, info) =
   | None   -> (e, Option.cata (Loc.add_loc info) info loc)
   | Some l -> (e, info)
 
-exception End_of_input
+let pr_time_header (vernac : vernac_control) =
+  (* Drop the time header to print the command, we should indeed use a
+     different mechanism to `-time` commands than the current hack of
+     adding a time control to the AST. *)
+  let vernac = match vernac with
+    | { v = { control = ControlTime _ :: control; attrs; expr }; loc } ->
+      CAst.make ?loc { control; attrs; expr }
+    | _ -> vernac
+  in
+  Topfmt.pr_cmd_header vernac
+
+let interp_control_flag ~vernac (f : control_flag) ~st
+    (fn : st:Vernacstate.t -> Vernacstate.LemmaStack.t option) =
+  match f with
+  | ControlFail ->
+    with_fail ~st (fun () -> fn ~st);
+    st.Vernacstate.lemmas
+  | ControlTimeout timeout ->
+    vernac_timeout ~timeout (fun () -> fn ~st) ()
+  | ControlTime batch ->
+    let header = if batch then pr_time_header vernac else Pp.mt () in
+    System.with_time ~batch ~header (fun () -> fn ~st) ()
```

Comment:
I don't like having to pass the vernac just for printing, it makes the type of interp_control_flag look misleading. Maybe we could pass the pre-printed header instead? (probably wrapped in lazy)

Reply:
Also maybe we should just print the expr without the attributes/controls. That change may be outside the scope of this PR though.

Reply:
Indeed I'd be afraid of the change as some users do rely on the printing to be exact, I dunno. I suggest we rename the argument for now?